### PR TITLE
Sending lazy attributes only when appropriate depending on the button interaction mode

### PR DESCRIPTION
**Pull request by gtorodelvalle**
- 100% tests passed
- Assigned to @dmoranj

### CODE REVIEW

File: lib/context_adapter_communication.js [hunk 1]
```diff
@@ -539,40 +593,94 @@ function notifyOperationCompleted(request, data, callback) {
...
+      request ? request.headers['fiware-service'] : thirdPartyUpdateRequest.headers['fiware-service'],
+      request ? request.headers['fiware-servicepath'] : thirdPartyUpdateRequest.headers['fiware-servicepath'],
+      payload,
+      callback
+    );
+  }
+
+  if (request) {
+    // Synchronous Third Party request
+    if (request.contextAdapter.operationDescriptor.interactionType ===
```
Comment (mrutid) [inline lib/context_adapter_communication.js]: are we sure that request contains a contextAdapter and an operationDescription?

**Comment by dmoranj** [comment]
There seem to be some random errors depending on the execution time:

```
  86 passing (4s)
  1 failing

  1) Context Adapter server: updateContext requests: Asynchronous button operation requests: Synchronous third party service: should notify the request as 'closed' if no Context Broker available to serve service information:
     Error: timeout of 2000ms exceeded
      at null.<anonymous> (/Users/dmj/Documents/Desarrollo/Projects/context-adapter/node_modules/mocha/lib/runnable.js:165:14)
      at Timer.listOnTimeout [as ontimeout] (timers.js:121:15)
```

It just happens in one execution out of ten or so, but we must be aware of this potential problem.

File: lib/context_adapter_communication.js [hunk 2]
```diff
@@ -539,40 +593,94 @@ function notifyOperationCompleted(request, data, callback) {
...
-  );
+    sendUpdateContext(
+      request ? request.headers['fiware-service'] : thirdPartyUpdateRequest.headers['fiware-service'],
+      request ? request.headers['fiware-servicepath'] : thirdPartyUpdateRequest.headers['fiware-servicepath'],
+      payload,
+      callback
+    );
+  }
+
+  if (request) {
```
Comment (dmoranj) [inline lib/context_adapter_communication.js]: This is a bit strange. I mean the use of a parameter to decide wether we are answering a syncrhonous or asynchronous request, inside this function. I think each branch of this if should be a sepparate function, invoking different `notifyOperationCompletedSync()` and `notifyOperationCompletedAsync()` functions depending on the type of call (and I assume you know what type of call you are doing, when you decide if you send a `request` object or a `null` value, in lines L182 and L276 of `context_adapter_server.js`). The common behaviour can be extracted to a decorator function that could be called in the asignation of the function to the exports attribute, i.e.: something like this

```
  getServiceDescriptor: getServiceDescriptor,
  sendThirdPartyRequest: sendThirdPartyRequest,
  notifyOperationCompletedSync: notifyOperationCompleted(doSyncOp),
  notifyOperationCompletedAsync: notifyOperationCompleted(doAsyncOp),
  notifyOperationClosed: notifyOperationClosed,
```

**Comment by dmoranj** [comment]
Just some general things to clean up the code a little bit (it seems a bit hard to read for me).

context_adapter_communications.js (L116) -> JSON Schemas belong to an external folder, and should be required in initialization time (the `require()` function can load JSON files as objects). And, if you need to validate requests, the proper place is as a middleware in the HTTP framework (it is currently mixed with the features of the CA).

context_adapter_communications.js (L349) -> The purpose of a function with the name `getButtonInteractionType()` seems fairly clear, but it gets obscure if the first parameter is the response you received in an earlier step, instead of the information the function needs to work. In this particular case, you just need two things of that object: the service and the service path, so the function would be clearer if it gets those two pieces of data as the only parameters. This same advice applies to other functions with the same problem. Changing the functions to work that way improves the maintainability and reusability of the code; i.e.: in this particular example, if you have another function that needs that kind of information, you could reuse the function as it is, just reading the description, without wondering what fields of the request must be filled in order for the function to work.

contet_adapter_communications.js (L429) -> Here you have three levels of functions (L433, L444, L451). This makes the code very difficult to read. You can move all the functions to an upper level, give them a name and a distinct purpose, and then use the named functions in the global code. Given a good name choice, reading the global code becomes much easier.

File: lib/context_adapter_communication.js (hunk 1 above)
Comment (gtorodelvalle) [inline lib/context_adapter_communication.js]: We (I am the one setting it) are... :p

File: lib/context_adapter_communication.js (hunk 2 above)
Comment (gtorodelvalle) [inline lib/context_adapter_communication.js]: Much much nicer :) Thanks!
Fixed in https://github.com/telefonicaid/context-adapter/commit/a8ec73809805cb953f592fa9e172d543e5937784

**Comment by gtorodelvalle** [comment]
As commented offline, I have just created https://github.com/telefonicaid/context-adapter/issues/7 to deal with the readability and maintainability suggestions which BTW I think are more than appropriate :)

I'll deal with them in a next PR not to block this one :)

On the other hand, I have created https://github.com/telefonicaid/context-adapter/issues/8 to track the issue related to the unit tests not passing randomly :)

Thanks!

**Comment by gtorodelvalle** [comment]
I have just deployed and confirmed that the happy path works perfectly for the 4 modes of operation (synchronous and asynchronous on the button and third party sides).

**Pull request merged**